docs/technical/git-submodules.md

```diff
@@ -0,0 +1,167 @@
+operation.
+
+## Roadmap
+
+Git submodules should be implemented in an order that supports an increasing set
+of workflows, with the goal of getting feedback early and often. When support is
+incomplete, jj should not crash, but instead provide fallback behavior and warn
+the user where needed.
+
+The goal is to land good support for pure Jujutsu repositories, while colocated
+repositories will be supported when convenient.
+
+This section should be treated as a set of guidelines, not a strict order of
+work.
+
+### Phase 1: Readonly submodules
+
+This includes work that inspects submodule contents but does not create new
+objects in the submodule. This requires a way to store submodules in a jj
+repository that supports readonly operations.
+
+#### Outcomes
+
+- Submodules can be cloned anew
+- New submodule commits can be fetched
+- Submodule history and branches can be viewed
+- Submodule contents are populated in the working copy
+- Superproject gitlink can be updated to an existing submodule commit
+- Conflicts in the superproject gitlink can be resolved to an existing submodule
+  commit
+
+### Phase 2: Snapshotting new changes
+
+This allows a user to write new contents to a submodule and its remote.
+
+#### Outcomes
+
+- Changes in the working copy can be recorded in a submodule commit
+- Submodule branches can be modified
+- Submodules and their branches can be pushed to their remote
+
+### Phase 3: Merging/rebasing/conflicts
+
+This allows merging and rebasing of superproject commits in a content-aware way
+(in contrast to Git, where only the gitlink commit ids are compared), as well as
+workflows that make resolving conflicts easy and sensible.
+
+This can be done in tandem with Phase 2, but will likely require a significant
+amount of design work on its own.
+
+#### Outcomes
+
+- Merged/rebased submodules result in merged/rebased working copy content
+- Merged/rebased working copy content can be committed, possibly by creating
+  sensible merged/rebased submodule commits
+- Merge/rebase between submodule and non-submodule gives a sensible result
+- Merge/rebase between submodule A and submodule B gives a sensible result
+
+### Phase ?: An ideal world
+
+I.e. outcomes we would like to see if there were no constraints whatsoever.
+
+- Rewriting submodule commits rewrites descendants correctly and updates
+  superproject gitlinks.
+- Submodule conflicts automatically resolve to the 'correct' submodule commits,
+  e.g. a merge between superproject commits creating a merge of the submodule
+  commits.
+- Nested submodules are as easy to work with as non-nested submodules.
+
+## Design
+
+### Guiding principles
+
+TODO
+
+### On-disk representation
+
+In the Git backend, submodules will be stored in Git's "absorbed submodule"
+format, i.e. a full Git repository with a Git directory in
+`<superproject-git-dir>/modules/<submodule-name>`, and a working tree with a
+`.git` file. Using Git's own format is the simplest option, even in a
```

Comment:
it seems odd to have a `.git` file in a tree that is managed by jj.

Reply:
it also creates a risk that people can run `git` in the jj submodule and create havoc by modifying things outside of JJ's knowledge.

Reply:
> I would also prefer not to have two different implementations for the jj-only and colocated cases. WDYT?

what is the long-term target? colocated sounds like it is an interim solution to ease the transition to a jj-only world?

Reply:
My long-term target is to support both colocated and jj-only repos.

> colocated sounds like it is an interim solution to ease the transition to a jj-only world?

I suppose that is true, but I am not so confident that the world will converge onto jj (as much as I would like that), or at least, not quickly enough. I expect that having a Git repo as a common denominator will be useful for a very very long time.